### PR TITLE
Fix extensions not working

### LINKED ISSUE #2926: Extensions not working
**Is this a request for help?**:
Yes
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE
---

**What version of acs-engine?**:
v0.16.2
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:
Running the extension examples provided doesn't work

**What you expected to happen**:
It worked

**How to reproduce it** (as minimally and precisely as possible):
Add the prometheus-grafana-k8s extension to deployment

**Anything else we need to know**:
After some fiddling I got the extension to be accepted in the ARM template, but then I instead got the error below during deployment:

```
{
  "error": {
    "code": "BadRequest",
    "message": "Multiple VMExtensions per handler not supported for OS type 'Linux'. VMExtension 'cse0' with handler 'Microsoft.Azure.Extensions.CustomScript' already added or specified in input."
  }
}
```

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2926 

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
@jackfrancis @ritazh @sozercan @trstringer @neurot1cal PTAL

**Comment by CecileRobertMichon** [comment]
@neurot1cal I think I found why you were running into an error with extensions even after that other PR. This PR should fix it.

File: pkg/acsengine/engine.go [hunk 1]
```diff
@@ -2854,10 +2854,10 @@ func internalGetPoolLinkedTemplateText(extTargetVMNamePrefix, orchestratorType,
 	if e != nil {
 		return "", e
 	}
-	if strings.Contains(extTargetVMNamePrefix, "agent") {
```
Comment (CecileRobertMichon) [inline pkg/acsengine/engine.go]: agent VMs do not contain "agent", only masters contain "master". this was causing the extension vm type to always be master.

File: extensions/choco/v1/template-link.json [hunk 2]
```diff
@@ -25,6 +25,9 @@
 			"targetVMName": {
 				"value": "[concat(EXTENSION_TARGET_VM_NAME_PREFIX, copyIndex(EXTENSION_LOOP_OFFSET))]"
 			},
+			"targetVMType": {
+                "value": "EXTENSION_TARGET_VM_TYPE"
```
Comment (jackfrancis) [inline extensions/choco/v1/template-link.json]: nit: whitespace indent

**Comment by huajunzhao** [comment]
Hi @CecileRobertMichon ..

Thanks for the fix.

It looks like the deployment is still failed. 

OPERATION ID7D0BCEDB12F43E24
TRACKING ID5b69691b-120d-4f72-8189-a9220448301e
STATUSConflict
PROVISIONING STATEFailed
TIMESTAMP5/23/2018, 10:53:39 AM
DURATION36 minutes 27 seconds
TYPEMicrosoft.Compute/virtualMachines/extensions
RESOURCE ID/subscriptions/58043194-91b4-4390-be55-dd64251a67bd/resourceGroups/kzha_k8s_rg/providers/Microsoft.Compute/virtualMachines/k8s-agentpool1-35463463-0/extensions/cse-agent-0
STATUSMESSAGE{
  "status": "Failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'Failed'.",
    "details": [
      {
        "code": "VMExtensionProvisioningError",
        "message": "VM has reported a failure when processing extension 'cse-agent-0'. Error message: \"Enable failed: failed to execute command: command terminated with exit status=4\"."
      }
    ]
  }
}
RESOURCEk8s-agentpool1-35463463-0/cse-agent-0

I'm using 0.17.1 acs-engine: https://github.com/Azure/acs-engine/releases/tag/v0.17.1 and also regenerate the arm template, but the deployment still has the same issue. 

Notes:  I was performing this on China Azure.

**Comment by neurot1cal** [comment]
@CecileRobertMichon The deployment went through successfully, no failed extension messages anymore! 

image

But when I look at the cluster, I dont actually see any prometheus/grafana pods. Any ideas?

image

**Comment by trstringer** [comment]
@neurot1cal did you deploy this on the master or agent node(s)?

**Comment by neurot1cal** [comment]
@trstringer agent. But I do have two linux pools, and one windows pool. It was playced in linuxpool2.

**Comment by neurot1cal** [comment]
@trstringer Same thing with a single linuxpool. Trying again with no windows pool;

```
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.10"
    },
    "masterProfile": {
      "count": 3,
      "dnsPrefix": "<REDACTED>",
      "vmSize": "Standard_D11_v2"
    },
    "agentPoolProfiles": [
      {
        "name": "linuxpool1",
        "count": 5,
        "vmSize": "Standard_D11_v2",
        "availabilityProfile": "AvailabilitySet",
        "storageProfile": "ManagedDisks",
        "extensions": [
          {
            "name": "prometheus-grafana-k8s"
          }
        ]
      }
    ],
    "linuxProfile": {
      "adminUsername": "<REDACTED>",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "<REDACTED>"
          }
        ]
      }
    },
    "extensionProfiles": [
      {
        "name": "prometheus-grafana-k8s",
        "version": "v1",
        "rootURL": "https://raw.githubusercontent.com/Azure/acs-engine/master/"
      }
    ],
    "servicePrincipalProfile": {
      "clientId": "<REDACTED>",
      "secret": "<REDACTED>"
    }
  }
}

```

**Comment by neurot1cal** [comment]
@trstringer @CecileRobertMichon Bingo. As soon as I removed the Windows nodes from the cluster definition, it worked. ;)

image

**Comment by trstringer** [comment]
We do not currently support clusters with Windows agents with this extension.

**Comment by neurot1cal** [comment]
@trstringer Understood - Adding to the list of things broken in with Windows nodes. :) 

Should `kubernetes-nodes-cadvisor` work out of the box? And is there a plan to have kubelets be discoverable as targets in Prometheus?

image

**Pull request merged**